addons/print_docsaway/models/print_docsaway.py

```diff
@@ -0,0 +1,391 @@
+                - Approved: y (successful transaction) or n (failed one)
+                - Price: price (in AUD)
+                - Reference: unique string to refer this transaction
+                - Date: date and time of transaction according to Docsaway
+                - Balance: return the remaining account balance after
+                    transaction (in AUD)
+        """
+        api_mode = 'LIVE' if self.environment == 'production' else 'TEST'
+        # build request
+        request = {
+            'APIMode': api_mode,
+            'Recipient': {
+                'name': address.get('name'),
+                'address1': address.get('street'),
+                'city': address.get('city'),
+                'zip': address.get('zip'),
+                'country': address.get('country_code'),
+            },
+            'PrintingStation': {
+                'id': printing_infos.get('station_id'),
+                'courierID': printing_infos.get('courier_id'),
+                'ink': printing_infos.get('ink', 'BW'),
+                'paper': printing_infos.get('paper', 80),
+            },
+            'PDFFile': PDFfile,
+        }
+        # complete facultative parameters
+        if address.get('street2'):
+            # avoid to return an error on facultative field
+            if len(address.get('street2')) <= 50:
+                request['Recipient'].update({
+                    'address2': address.get('street2'),
+                })
+        if address.get('state'):
+            request['Recipient'].update({
+                'state': address.get('state'),
+            })
+        return self._docsaway_do_server_request(_DOCSAWAY_SENDMAIL_URL, request, _DOCSAWAY_ERROR_MAIL)
+
+
+class print_order(models.Model):
+
+    _inherit = 'print.order'
+
+    # station fields
+    docsaway_station = fields.Char('Docsaway Station', help="Docsaway Station ID.")
+    docsaway_courier = fields.Char('Docsaway Courier', help="Docsaway Printer ID.")
+    docsaway_zone = fields.Integer('Docsaway Zone', help="Docsaway Station ID.")
+    # transaction fields
+    docsaway_reference = fields.Char('Docsaway Reference', help="Docsaway Reference Order.")
+    docsaway_approved = fields.Char('Docsaway Approved', help="Docsaway Transation Status.")
+
+
+    def _docsaway_prepare_printing(self):
+        """ Pre-process the sending by computing the best station for each given orders.
+        """
+        # minimize the number of call to DocsAway API by grouping by partner
+        # (for localization) and by ink to find the best station.
+        stations = {}
+        for order in self:
+            try:
+                if not (order.partner_id.id, order.ink) in stations:
+                    stations[(order.partner_id.id, order.ink)] = order.provider_id._docsaway_fetch_station(order.partner_id.country_id.code, order.ink, order.paper_weight)
+                response = stations[(order.partner_id.id, order.ink)]
+                order.write({
```

Comment:
I would move the write outside of the try / except statement, as it should not be in the same protection.